### ISSUE TITLE
[FEATURE REQUEST] Reuse existing EFS file systems in different docker volumes

### ISSUE THREAD

**Issue by bvis** [codedellemc/rexray#698]
### Fact
Currently there's a basic support of EFS as filesystem provider for Rex-Ray. But as far as I've seen it creates/uses a filesystem for each docker volume.

As the EFS paradigm (unlimited size, credits/performance model, limits) is quite different from other providers it would have sense to add the possibility to define in configuration one particular EFS backend for the node. 

### Proposal

Instead of the current configuration that is something like:

```
libstorage:
 service: efs
 server:
   services:
     efs:
       driver: efs
       efs:
         accessKey:      ${AWS_ACCESS_KEY_ID}
         secretKey:      ${AWS_SECRET_ACCESS_KEY}
         securityGroups: ${AWS_SECURITY_GROUP_ID}
         region:         ${AWS_DEFAULT_REGION}
         tag:            efs-namespace
```
Would be good to have another extra configuration value:
```
libstorage:
 service: efs
 server:
   services:
     efs:
       driver: efs
       efs:
         accessKey:      ${AWS_ACCESS_KEY_ID}
         secretKey:      ${AWS_SECRET_ACCESS_KEY}
         securityGroups: ${AWS_SECURITY_GROUP_ID}
         region:         ${AWS_DEFAULT_REGION}
         tag:            efs-namespace
         fileSystem:     concrete-fs
```
The usage in both scenarios would be the same:
```
$ docker run -d -P \
  --volume-driver=rexray \
  -v my-named-volume:webapp \
  --name web \
  user/app-image

$ docker run -d -P \
  --volume-driver=rexray \
  -v my-named-volume:db \
  --name db \
  mysql
```
The only difference is that in the first configuration (currently available) you are creating or using a volume that represents each EFS filesystem: *efs-namespace:webapp* and *efs-namespace:db*. Both mounting the root folder of each fs.

The second allows to reuse a global EFS filesystem and mount just a directory from it. Applying the same definitions it would be in EFS just one file system: *efs-namespace:concrete-fs*, but you would be mounting a subdirectory: *efs-namespace:concrete-fs/webapp* and *efs-namespace:concrete-fs/db*

### Advantages

- Limitation of number of EFS filesystem per AWS account is no longer a problem
- Credit usage: EFS performs better with the usage of the FS. Divide the usage on each independent FS and you'll dividing individual performance as well.

**Comment by akutz** [comment]
Hi @bvis,

This is a fantastic idea! In practice however there will be a few issues all stemming from the fact that libStorage is designed to be "volume"-centric where "volume" is an atomic storage container. In many cases that's a block-based volume, but for NFS-based storage it's an exported file system. However, because block is the basis for much of the design, the concept of further dividing up the atomic storage endpoint is just not supported. It doesn't mean it's not a good idea. Like I said, it's in fact a fantastic one. Much like another user who wished we could connect to VMDK files that might provide multiple file systems, the ultimate storage endpoint currently has to be atomic.

If the only change was in the driver itself, I'd be all for working on this or suggesting you fork and work on this. But there would need to be changes in the Docker integration driver to account for avoiding NFS remounts as well. And that's off the top of my head. I'm sure there will be a few more framework changes to account for this request.

Does my response make sense? I hope I conveyed the reasons that this isn't a piece of low-hanging fruit as perhaps you thought it might have been. Thank you!

**Comment by bvis** [comment]
@akutz Thanks for your quick and detailed answer. It has all the sense of the world.

**Comment by codenrhoden** [comment]
Issue moved to [codedellemc/libstorage #468](https://github.com/codedellemc/libstorage/issues/468) via [**ZenHub**](https://www.zenhub.com/)

**Comment by ryanmickler** [comment]
What happened with this issue, i think the link is broken?

**Comment by ryanmickler** [comment]
I'm willing to do some work on this. Perhaps we can add an option `mount_path=/data1/` etc

**Comment by clintkitson** [comment]
@ryanmickler It is possible that this is actually already available through the `linux.volume.rootpath`  parameter. This currently maps the `/data` directory, you can be used to mount any subdirectory of choice. Does that make sense?

**Comment by ryanmickler** [comment]
Yeah I had a look at this option to figure out if it does what i thought.
If this really just does modify the "Source" in the mount info (form `docket inspect`) as in`"/mnt/docker/plugins/5337c23c0cb4b521a4eb7718f60f555e10df3d7af0aeaca1c54c58b38c5987cb/rootfs/var/lib/rexray/volumes/myvolume/{linux.volume.rootpath}"`
then it would be great.

**Comment by clintkitson** [comment]
That is exactly it. The path you specified gets passed to docker.

**Comment by ryanmickler** [comment]
I can't seem to get it to work, what shoudl the syntax be?
`-opt linux.volume.rootpath="blah"`

**Comment by clintkitson** [comment]
That's a service level configurable feature unfortunately and not applied
individually per volume.

**Comment by ryanmickler** [comment]
Yeah that was my primary concern with this approach, however, i only need to mount one volume directory per host, so it will still fit my use case. Still can't get it to work though

**Comment by ryanmickler** [comment]
From looking at the source, it seems to have changed to 
`libstorage.integration.volume.operations.mount.rootpath`

**Comment by clintkitson** [comment]
Try this one. That is the fully path, but linux.volume.rootPath is an alias to it. You can also run `rexray env | grep root` to see configured values.

```
libstorage:
 service: efs
 server:
   services:
     efs:
       driver: efs
       efs:
         accessKey:      ${AWS_ACCESS_KEY_ID}
         secretKey:      ${AWS_SECRET_ACCESS_KEY}
         securityGroups: ${AWS_SECURITY_GROUP_ID}
         region:         ${AWS_DEFAULT_REGION}
         tag:            efs-namespace
linux:
  volume:
    rootPath: /testing/test
```

**Comment by ryanmickler** [comment]
So as far as i can tell, this is more of a problem with the docker syntax for mounting named volumes. argh. i'll go complain on the moby thread too.

**Comment by SPhu** [comment]
+1 
Just want to say this is a good use case.  If you can set a sourcepath for local volumes, why cant you do it for rexray?
The only way around it for me right now it just not using rexray and manually setting up a single nfs mount on the host
